aries/aries_vcx_core/src/wallet/askar/askar_did_wallet.rs

```diff
@@ -0,0 +1,178 @@
+    rng_method::RngMethod,
+    sig_type::SigType,
+    AskarWallet,
+};
+use crate::{
+    errors::error::{AriesVcxCoreError, AriesVcxCoreErrorKind, VcxCoreResult},
+    wallet::{
+        base_wallet::{did_data::DidData, DidWallet},
+        constants::{DID_CATEGORY, TMP_DID_CATEGORY},
+        structs_io::UnpackMessageOutput,
+        utils::did_from_key,
+    },
+};
+
+#[async_trait]
+impl DidWallet for AskarWallet {
+    async fn create_and_store_my_did(
+        &self,
+        seed: Option<&str>,
+        _did_method_name: Option<&str>,
+    ) -> VcxCoreResult<DidData> {
+        let mut tx = self.backend.transaction(self.profile.clone()).await?;
+        let (did, local_key) = self
+            .insert_key(
+                &mut tx,
+                KeyAlg::Ed25519,
+                seed_from_opt(seed).as_bytes(),
+                RngMethod::RandomDet,
+            )
+            .await?;
+        let verkey = local_key_to_bs58_public_key(&local_key)?;
+        self.insert_did(&mut tx, &did, DID_CATEGORY, &verkey, None)
+            .await?;
+        tx.commit().await?;
+        Ok(DidData::new(&did, key_from_base58(&verkey)?))
+    }
+
+    async fn key_for_did(&self, did: &str) -> VcxCoreResult<Key> {
+        let data = self
+            .find_current_did(&mut self.backend.session(self.profile.clone()).await?, did)
+            .await?;
+
+        if let Some(did_data) = data {
+            Ok(did_data.verkey().to_owned())
+        } else {
+            Err(AriesVcxCoreError::from_msg(
+                AriesVcxCoreErrorKind::WalletRecordNotFound,
+                format!("did not found in wallet: {}", did),
+            ))
+        }
+    }
+
+    async fn replace_did_key_start(&self, did: &str, seed: Option<&str>) -> VcxCoreResult<Key> {
+        let mut tx = self.backend.transaction(self.profile.clone()).await?;
+        if self.find_current_did(&mut tx, did).await?.is_some() {
+            let (new_key_name, local_key) = self
+                .insert_key(
+                    &mut tx,
+                    KeyAlg::Ed25519,
+                    seed_from_opt(seed).as_bytes(),
+                    RngMethod::RandomDet,
+                )
+                .await?;
+            let verkey = local_key_to_bs58_public_key(&local_key)?;
+            self.insert_did(&mut tx, did, TMP_DID_CATEGORY, &verkey, None)
+                .await?;
+            tx.commit().await?;
+            key_from_base58(&new_key_name)
+        } else {
+            Err(AriesVcxCoreError::from_msg(
+                AriesVcxCoreErrorKind::WalletRecordNotFound,
+                format!("did not found in wallet: {}", did),
+            ))
+        }
+    }
+
+    async fn replace_did_key_apply(&self, did: &str) -> VcxCoreResult<()> {
+        let mut tx = self.backend.transaction(self.profile.clone()).await?;
+        if let Some(did_data) = self.find_did(&mut tx, did, TMP_DID_CATEGORY).await? {
+            let verkey_did = did_data.did_from_verkey();
```

Comment:
At this point, `did` must equal to the value of the variable `verkey_did`, do you agree? If it's not, something must be had been done prior to this. This is contract expected to hold - it points me more fundamentally to the fact that there seem to be redundancy in how the data is being stored - it seem that it is pair of :
```
did  ----> DidData { did, verkey }
```
which is what brings us on this line into "dilemma", what's the source of truth of what `did` we work with at this point, and we have 3 options in fact:
- arg `did`
- `did_data.did`
-  derive the did from `did_data.verkey` (which is what is being done now)

When we pass in our temporary did as argument to this function, we only need it to resolve the full value of verkey. That should be the only purpose of the the `find_did` lookup here.

Reply:
> At this point, did must equal to the value of the variable verkey_did, do you agree?
Yes.

I removed `did_data.verkey` as it is not needed. It seems like the did itself does not need to be stored as a part of the record value, it is just something the existing wallet does - removed as well.

Reply:
This was rolled back. `did_data.verkey` for the did/key names to correctly connect when rotating key more than once.